cpp/src/arrow/dataset/discovery.h

```diff
@@ -85,35 +85,97 @@ class ARROW_DS_EXPORT DataSourceDiscovery {
   std::shared_ptr<Expression> root_partition_;
 };
 
-/// \brief FileSystemDataSourceFactory creates a DataSource from a vector
-/// of fs::FileStats or a fs::Selector.
+struct FileSystemDiscoveryOptions {
+  // For the purposes of applying the partition scheme, paths will be stripped
+  // of the partition_base_dir. Files not matching the partition_base_dir
+  // prefix will be skipped for partition discovery. The ignored files will still
+  // be part of the DataSource, but will not have partition information.
+  //
+  // Example:
+  // partition_base_dir = "/dataset";
+  //
+  // - "/dataset/US/sales.csv" -> "US/sales.csv" will be given to the partition
+  //                              scheme
+  //
+  // - "/home/john/late_sales.csv" -> Will be ignored for partition discovery.
+  //
+  // This is useful for partition schemes which parses directory when ordering
+  // is important, e.g. SchemaPartitionScheme.
+  std::string partition_base_dir;
+
+  // Files given (via selector or explicitly) will be filtered via the
+  // FileFormat::IsSupported method.  This will incur IO for each files in a
+  // serial and single threaded fashion. Disabling this feature will skip the
+  // IO, but unsupported files may will be present in the DataSource
+  // (resulting in an error at scan time).
+  bool filter_supported_files = true;
+
+  // Files matching one of the following prefix will be ignored by the
+  // discovery process. This is matched to the basename of a path.
+  //
+  // Example:
+  // ignore_prefixes = {"_", ".DS_STORE" };
+  //
+  // - "/dataset/data.csv" -> not ignored
+  // - "/dataset/_metadata" -> ignored
+  // - "/dataset/.DS_STORE" -> ignored
+  std::vector<std::string> ignore_prefixes = {
```

Comment:
Why just prefixes? Seems like more of a `.gitignore` model would make sense. (Feel free to defer as scope creep.)

Reply:
if we reopened https://issues.apache.org/jira/browse/ARROW-6615 or https://issues.apache.org/jira/browse/ARROW-6257 then we could refactor `ignore_prefixes` to `ignore_globs`. Alternatively, these could be `std::regex`. In any case I do think this behavior should be addressed in follow up

Reply:
I think regex would be the easiest to integrate with other languages, we only need  to fix the toolchain issue. I can't recall which compiler doesn't support it, mingw?

Reply:
GCC 4.8.2, IIRC. See for example: https://issues.apache.org/jira/browse/ARROW-6906
I believe the dataset module is disabled for the builds which use that, however